### PR TITLE
Add support for more straight forward XGBoost API

### LINKED ISSUE #1829: More straight forward XGBoost API
Instead of writting

``` sql
SELECT * FROM titanic.train
TO TRAIN xgboost.gbtree 
WITH objective="binary:logistic"
LABEL survived
INTO sqlflow_models.my_xgboost_model;
```

the following naming might be more straight forward
``` sql
SELECT * FROM titanic.train
TO TRAIN XGBOOST.BINARYCLASSIFIER
LABEL survived
INTO sqlflow_models.my_xgboost_model;
```

I feel `XGBOOST.BINARYCLASSIFIER` is a good fit for the first SQLFlow statement since it doesn't have the WITH clause. @llxxxll What do you think?

credit: This definition is original proposed by @sperlingxx at https://github.com/sql-machine-learning/sqlflow/blob/4f08b7cb8fc88269c8d0066cca8992a0b4658f7f/sql/codegen_ant_xgboost.go#L575

### PR DESCRIPTION
**Pull request by tonyyang-svail**
Fix #1829

### CODE REVIEW

**Comment by shendiaomo** [comment]
Should we use the sklearn API as described in https://xgboost.readthedocs.io/en/latest/python/python_api.html#module-xgboost.sklearn?
The users may have a lower migration cost.

**Comment by tonyyang-svail** [comment]
@shendiaomo I was thinking about the similar idea. However, I haven't find sklearn API example with SHAP https://github.com/sql-machine-learning/sqlflow/issues/749#issuecomment-527315312, so maybe it is better to stay on the native booster API.

**Comment by tonyyang-svail** [comment]
@shendiaomo btw, the [implementation](https://github.com/dmlc/xgboost/blob/71a8b8c65afcad6a3e3bf48bf1ee5e66d762543f/python-package/xgboost/sklearn.py#L1009-L1012) of `XGBRegressor` is nothing but specifying the objective.

```python
@xgboost_model_doc(
    "Implementation of the scikit-learn API for XGBoost regression.",
    ['estimators', 'model', 'objective'])
class XGBRegressor(XGBModel, XGBRegressorBase):
    # pylint: disable=missing-docstring
    def __init__(self, objective="reg:squarederror", **kwargs):
        super().__init__(objective=objective, **kwargs)
```

**Comment by shendiaomo** [comment]
> @shendiaomo btw, the [implementation](https://github.com/dmlc/xgboost/blob/71a8b8c65afcad6a3e3bf48bf1ee5e66d762543f/python-package/xgboost/sklearn.py#L1009-L1012) of `XGBRegressor` is nothing but specifying the objective.
> 
> ```python
> @xgboost_model_doc(
>     "Implementation of the scikit-learn API for XGBoost regression.",
>     ['estimators', 'model', 'objective'])
> class XGBRegressor(XGBModel, XGBRegressorBase):
>     # pylint: disable=missing-docstring
>     def __init__(self, objective="reg:squarederror", **kwargs):
>         super().__init__(objective=objective, **kwargs)
> ```

My point is that we can try to keep the names being consistent with the sklearn API, so users familiar with the sklearn classes can easily switch to their SQLFlow counterparts.

**Comment by tonyyang-svail** [comment]
> My point is that we can try to keep the names being consistent with the sklearn API, so users familiar with the sklearn classes can easily switch to their SQLFlow counterparts.

@shendiaomo Thanks for mentioning this. By names, do you mean the model name or parameter name? 

If you mean parameter names, I believe the parameter set of sklearn is a subset of native booster's, except one parameter `n_jobs`.

If you mean model names, do you want `XGBRegressor`, `xgboost.XGBRegressor`, or `xgboost.regressor`?

**Comment by shendiaomo** [comment]
> > My point is that we can try to keep the names being consistent with the sklearn API, so users familiar with the sklearn classes can easily switch to their SQLFlow counterparts.
> 
> @shendiaomo Thanks for mentioning this. By names, do you mean the model name or parameter name?
> 
> If you mean parameter names, I believe the parameter set of sklearn is a subset of native booster's, except one parameter `n_jobs`.
> 
> If you mean model names, do you want `XGBRegressor`, `xgboost.XGBRegressor`, or `xgboost.regressor`?

IMO `XGBRegressor` is better, there're mainly 5 classes in the sklearn API, i.e. `XGBClassifier`,  `XGBRFClassifier`, `XGBRFRegressor`, `XGBRanker`, `XGBRegressor`. Supporting all of them is great if we want to keep consistent with the user experience of the sklearn API.

File: doc/model_parameter.md [hunk 1]
```diff
@@ -10,16 +10,19 @@ SQLFlow connects a SQL engine (e.g., MySQL, Hive, or MaxCompute) and TensorFlow
...
     train.num_boost_round = 30,
     validation.select = "SELECT * FROM boston.val LIMIT 8"
-COLUMN crim, zn, indus, chas, nox, rm, age, dis, rad, tax, ptratio, b, lstat
 LABEL medv
 INTO sqlflow_models.my_xgb_regression_model;
 ```
 
+#### Model Types
+
+`XGBOOST.XGBREGRESSOR`, `XGBOOST.XGBCLASSIFIER`, `XGBOOST.XGBRFCLASSIFIER`, `XGBOOST.XGBRANKER`, `XGBOOST.GBTREE`, `XGBOOST.GBLINEAR`, `XGBOOST.DART`
```
Comment (typhoonzero) [inline doc/model_parameter.md:24]: Also add types for binary classification?

Comment (shendiaomo) [reply doc/model_parameter.md:24]: In the `sklearn` API, `XGBClassifier` defaults to binary classification.

Comment (typhoonzero) [reply doc/model_parameter.md:24]: Seems current implement only use `XGBOOST.XGBCLASSIFIER` with `multi:softprob`, how to use `binary:logistic` with the new model types?

Comment (tonyyang-svail) [reply doc/model_parameter.md:24]: @typhoonzero correct me if I am wrong. I believe the softmax loss and logistic loss are equivalent when dealing with binary classification. https://stats.stackexchange.com/a/254071

Comment (typhoonzero) [reply doc/model_parameter.md:24]: It's true. The difference is that if always use `multi:softprob` you always have to specify `num_classes` parameter. Meanwhile, I didn't test if https://github.com/dmlc/xgboost/issues/4853 occurs now.

Comment (tonyyang-svail) [reply doc/model_parameter.md:24]: @typhoonzero @shendiaomo Thanks for helping out. I will change `XGBClassifier`'s `objective` default to `binary:logistic`.

**Review by typhoonzero — APPROVED** [review]
